modules/tinymce/src/core/main/ts/paste/Clipboard.ts

```diff
@@ -0,0 +1,316 @@
+    const clipboardContent = getDataTransferItems(e.clipboardData);
+
+    if (!hasHtmlOrText(clipboardContent) && pasteImageData(editor, e, getLastRng())) {
+      return;
+    }
+
+    // If the clipboard API has HTML then use that directly
+    if (hasContentType(clipboardContent, 'text/html')) {
+      e.preventDefault();
+      insertClipboardContent(editor, clipboardContent, clipboardContent['text/html'], plainTextMode);
+    } else {
+      // We can't extract the HTML content from the clipboard so we need to allow the paste
+      // to run via the pastebin and then extract from there
+      pasteBin.create();
+      Delay.setEditorTimeout(editor, () => {
+        // Get the pastebin content and then remove it so the selection is restored
+        const html = pasteBin.getHtml();
+        pasteBin.remove();
+        insertClipboardContent(editor, clipboardContent, html, plainTextMode);
+      }, 0);
+    }
+  });
+};
+
+const registerDataImageFilter = (editor: Editor) => {
+  const isWebKitFakeUrl = (src: string): boolean => Strings.startsWith(src, 'webkit-fake-url');
+  const isDataUri = (src: string): boolean => Strings.startsWith(src, 'data:');
+  const isPasteInsert = (args: ParserArgs): boolean => args.data?.paste === true;
+
+  // Remove all data images from paste for example from Gecko
+  // except internal images like video elements
+  editor.parser.addNodeFilter('img', (nodes, name, args) => {
+    if (!Options.shouldPasteDataImages(editor) && isPasteInsert(args)) {
+      for (const node of nodes) {
+        const src = node.attr('src');
+        if (Type.isString(src) && !node.attr('data-mce-object') && src !== Env.transparentSrc) {
+          // Safari on Mac produces webkit-fake-url see: https://bugs.webkit.org/show_bug.cgi?id=49141
+          if (isWebKitFakeUrl(src)) {
+            node.remove();
+          } else if (!Options.shouldAllowHtmlDataUrls(editor) && isDataUri(src)) {
+            node.remove();
+          }
+        }
+      }
+    }
+  });
+};
+
+/*
+ * This class contains logic for getting HTML contents out of the clipboard.
+ *
+ * We need to make a lot of ugly hacks to get the contents out of the clipboard since
+ * the W3C Clipboard API is broken in all browsers that have it: Gecko/WebKit/Blink.
+ * We might rewrite this the way those API:s stabilize. Browsers doesn't handle pasting
+ * from applications like Word the same way as it does when pasting into a contentEditable area
+ * so we need to do lots of extra work to try to get to this clipboard data.
+ *
+ * Current implementation steps:
+ *  1. On keydown with paste keys Ctrl+V or Shift+Insert create
+ *     a paste bin element and move focus to that element.
+ *  2. Wait for the browser to fire a "paste" event and get the contents out of the paste bin.
+ *  3. Check if the paste was successful if true, process the HTML.
+ *  (4). If the paste was unsuccessful use IE execCommand, Clipboard API, document.dataTransfer old WebKit API etc.
+ */
```

Comment:
This seems out of date and in need of some minor corrections :)

Reply:
I knew I should have read this massive wall of text 😢  anyways have reworded this significantly, so how does that look?

Reply:
much better 👍